lib/editor/components/pattern/TripPatternViewer.js

```diff
@@ -35,6 +35,9 @@ export default class TripPatternViewer extends Component<Props> {
       updateActiveGtfsEntity
     } = this.props
     if (!activePattern) return null
+    // FLEX TODO: Is this the best place to put it?
+    activePattern.patternLocations = activePattern.patternLocations.map(pl => { return {...pl, id: `${pl.locationId}${pl.stopSequence}`} })
+    activePattern.patternStops = activePattern.patternStops.map(ps => { return {...ps, id: `${ps.stopId}${ps.stopSequence}`} })
```

Comment:
The render method should not change props. You can clone, or precompute this stuff in a state variable that gets updated when `componentDidUpdate` occurs.

Reply:
I made this change, but had to revert it because the state change caused a re-render which turned out to be really expensive because all the child components also re-rendered (I think this could've been avoided by moving to a newer version of react). 

We can't pre-compute them either, since the props can change.

Not sure if you have a better solution beyond re-architecting the entire trip pattern viewer.

Reply:
Just tried clone, but this also caused strange behavior (the popups broke) -- my guess is that there is some weird hacks going on elsewhere

Reply:
You may need to clone `this.props` and modify the active pattern in there, then pass the cloned `props` to the other components rendered, so that all use the same data.